core/jazz_acl/components/scm/gitlab.js

```diff
@@ -0,0 +1,213 @@
+      let repoInfo = {
+        "permission": config.ACCESS_LEVEL[policy.permission],
+        "gitlabRepoId": repo_id
+      };
+
+      if (_.includes(config.PERMISSION_CATEGORIES, policy.category)) {
+        let gitlab_useId = await getGitlabUserId(config, policy.userId);
+        repoInfo.gitlabUserId = gitlab_useId;
+        let member_res = await getGitlabProjectMember(config, repoInfo);
+        let res;
+        if (member_res.isMember)
+          res = await updatePermsInGitlabRepo(config, repoInfo);
+        else
+          res = await addPermsInGitlabRepo(config, repoInfo);
+        if (res)
+          users_list.push(policy.userId);
+
+        count++
+        if (count < policies.details.length)
+          await addUserPermsInGitlabRepo(count);
+        else
+          return;
+
+      } else {
+        count++
+        if (count < policies.details.length)
+          await addUserPermsInGitlabRepo(count);
+        else
+          return;
+      }
+    } catch (ex) {
+      count++
+      if (count < policies.details.length)
+        await addUserPermsInGitlabRepo(count);
+      else
+        return;
+    }
+  }
+
+  await addUserPermsInGitlabRepo(0);
+  return users_list;
+};
+
+let addPermsInGitlabRepo = (config, repoInfo) => {
+  return new Promise((resolve, reject) => {
+    let dataString = `user_id=${repoInfo.gitlabUserId}&access_level=${repoInfo.permission}`;
+    let payload = {
+      url: `${config.HOSTNAME}${config.REPO_BASE_API}${repoInfo.gitlabRepoId}/members/`,
+      method: 'POST',
+      body: dataString,
+      rejectUnauthorized: false,
+      headers: {
+        'Accept': 'application/json',
+        'Accept-Charset': 'utf-8',
+        'Private-Token': config.PRIVATE_TOKEN
+      }
+    };
+    request(payload, (error, response, body) => {
+      if (error)
+        return reject(error);
+      else {
+        if (response.statusCode != 200 && response.statusCode != 201) {
+          return reject({ "error": "Error in addPermsInGitlabRepo." });
+        }
+        return resolve(response);
+      }
+    });
+  });
+}
+
+let updatePermsInGitlabRepo = (config, repoInfo) => {
+  return new Promise((resolve, reject) => {
+    let payload = {
+      url: `${config.HOSTNAME}${config.REPO_BASE_API}${repoInfo.gitlabRepoId}/members/${repoInfo.gitlabUserId}/?access_level=${repoInfo.permission}`,
+      method: 'PUT',
+      rejectUnauthorized: false,
+      headers: {
+        'Accept': 'application/json',
+        'Accept-Charset': 'utf-8',
+        'Private-Token': config.PRIVATE_TOKEN
+      }
+    };
+
+    request(payload, (error, response, body) => {
+      if (error)
+        return reject(error);
+      else {
+        if (response.statusCode != 200 && response.statusCode != 201) {
+          return reject({ "error": "Error in updatePermsInGitlabRepo." });
+        }
+        return resolve(response);
+      }
+    });
+  });
+}
+
+let getGitlabProjectMember = (config, repoInfo) => {
```

Comment:
you shld use async here wherenver u r using promise

Reply:
Not really required as discussed